File: android/app/src/main/java/com/antlib/hingewave/core/Spring.kt
package com.antlib.hingewave.core

import kotlin.math.PI
import kotlin.math.exp

/**
 * Critically damped spring with an exact closed-form step.
 * Same formula as core/reference/hingewave_ref/spring.py.
 */
class Spring(hz: Double) {
    val omega = 2.0 * PI * hz
    var x = 0.0
        private set
    var v = 0.0
        private set

    fun reset(value: Double) {
        x = value
        v = 0.0
    }

    /** Advances toward [target] by [dt] seconds and returns the new position. */
    fun step(target: Double, dt: Double): Double {
        if (dt <= 0.0) return x
        val w = omega
        val d = x - target
        val e = exp(-w * dt)
        val xNew = target + (d + (v + w * d) * dt) * e
        val vNew = (v - w * (v + w * d) * dt) * e
        x = xNew
        v = vNew
        return x
    }
}

File: windows/Hingewave.Core/Spring.cs
namespace Hingewave.Core;

/// <summary>
/// Critically damped spring with an exact closed-form step.
/// Same formula as core/reference/hingewave_ref/spring.py.
/// </summary>
public sealed class Spring
{
    public double Omega { get; }
    public double X { get; private set; }
    public double V { get; private set; }

    public Spring(double hz)
    {
        Omega = 2.0 * Math.PI * hz;
    }

    public void Reset(double value)
    {
        X = value;
        V = 0.0;
    }

    /// <summary>Advances toward <paramref name="target"/> by <paramref name="dt"/> seconds.</summary>
    public (double X, double V) Step(double target, double dt)
    {
        if (dt <= 0.0) return (X, V);
        var w = Omega;
        var d = X - target;
        var e = Math.Exp(-w * dt);
        var xNew = target + (d + (V + w * d) * dt) * e;
        var vNew = (V - w * (V + w * d) * dt) * e;
        X = xNew;
        V = vNew;
        return (X, V);
    }
}
